Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -437,22 +448,28 @@ class BrowserViewController: UIViewController {
     }
     
     fileprivate func setupTabs() {
-        let isPrivate = Preferences.Privacy.privateBrowsingOnly.value
-        let noTabsAdded = tabManager.tabsForCurrentMode.isEmpty
-        
-        var tabToSelect: Tab?
-        
-        if noTabsAdded {
-            // Two scenarios if there are no tabs in tabmanager:
-            // 1. We have not restored tabs yet, attempt to restore or make a new tab if there is nothing.
-            // 2. We are in private browsing mode and need to add a new private tab.
-            tabToSelect = isPrivate ? tabManager.addTab(isPrivate: true) : tabManager.restoreAllTabs()
-        } else {
-            tabToSelect = tabManager.tabsForCurrentMode.last
-        }
-        
         contentBlockListDeferred?.uponQueue(.main) { _ in
-            self.tabManager.selectTab(tabToSelect)
+            let isPrivate = Preferences.Privacy.privateBrowsingOnly.value
+            let noTabsAdded = self.tabManager.tabsForCurrentMode.isEmpty
+            
+            var tabToSelect: Tab?
+            
+            if noTabsAdded {
+                // Two scenarios if there are no tabs in tabmanager:
+                // 1. We have not restored tabs yet, attempt to restore or make a new tab if there is nothing.
+                // 2. We are in private browsing mode and need to add a new private tab.
+                tabToSelect = isPrivate ? self.tabManager.addTab(isPrivate: true) : self.tabManager.restoreAllTabs()
+            } else {
+                tabToSelect = self.tabManager.tabsForCurrentMode.last
+            }
+            if self.taskQueue.isEmpty {
```

Comment:
This is a bit more complex with promises, since you can't (AFAIK), identify whether a promise has a closure attached to it, so we probably need to handle this with additional metadata:

```
private class onLoad {
    fileprivate let await = Deferred<Void>()
    fileprivate var containsTasks = false
    fileprivate func upon(_ block: @escaping () -> Void) {
        await.upon(block)
        containsTasks = true
    }
}
```

So then here we'd do something like:
`if onLoad.containsTasks {`